File: adapter/src/main/java/com/iheartradio/multitypeadapter/GridLayoutSpanHandler.java
package com.iheartradio.multitypeadapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import com.iheartradio.multitypeadapter.interfaces.SpanHandler;

/**
 * Default implementation of SpanHandler that uses the TypeAdapters
 * getSpan() method to calculate spans for each item.
 *
 * This is only utilized for GridLayoutManagers.
 */
public class GridLayoutSpanHandler implements SpanHandler {

    private final MultiTypeAdapter mAdapter;
    private boolean mSpanLookupSet;

    GridLayoutSpanHandler(@NonNull final MultiTypeAdapter adapter){
        mAdapter = adapter;
        mSpanLookupSet = false;
    }

    @Override
    public void calculateSpan(@NonNull final RecyclerView.LayoutManager layoutManager) {
        if (!mSpanLookupSet && layoutManager instanceof GridLayoutManager) {
            ((GridLayoutManager)layoutManager).setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    return mAdapter.getTypeAdapterForPosition(position).getSpan();
                }
            });
            mSpanLookupSet = true;
        }
    }
}

File: adapter/src/main/java/com/iheartradio/multitypeadapter/GridLayoutSpanHandler.kt
package com.iheartradio.multitypeadapter

import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView

/**
 * Default implementation of SpanHandler that uses the TypeAdapters
 * getSpan() method to calculate spans for each item.
 *
 * This is only utilized for GridLayoutManagers.
 */
class GridLayoutSpanHandler internal constructor(private val adapter: MultiTypeAdapter) : SpanHandler {

    private var spanLookupSet: Boolean = false

    override fun calculateSpan(layoutManager: RecyclerView.LayoutManager) {
        if (!spanLookupSet && layoutManager is GridLayoutManager) {
            layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    return adapter.getTypeAdapterForPosition(position).span
                }
            }
            spanLookupSet = true
        }
    }
}
